File: app/src/main/java/com/pranav/java/ide/adapter/ProjectAdapter.java
package com.pranav.java.ide.adapter;

import android.view.View;
import android.view.ViewGroup;
import android.view.LayoutInflater; 
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import androidx.recyclerview.widget.DiffUtil;

import com.pranav.java.ide.R;
import com.pranav.project.mode.JavaProject;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

public class ProjectAdapter extends RecyclerView.Adapter<ProjectAdapter.ViewHolder> {

    public interface OnProjectSelectedListener {
        void onProjectSelect(JavaProject project);
    }

    public interface OnProjectLongClickedListener {
        boolean onLongClicked(View view, JavaProject project);
    }

    private final List<JavaProject> mProjects = new ArrayList<>();
    private OnProjectSelectedListener onProjectSelectedListener;
    private OnProjectLongClickedListener onProjectLongClickedListener; 

    public ProjectAdapter() {}

    public void setOnProjectSelectedListener(OnProjectSelectedListener onProjectSelectedListener) {
        this.onProjectSelectedListener = onProjectSelectedListener;
    }
    
    public void setOnProjectLongClickedListener(OnProjectLongClickedListener onProjectLongClickedListener) {
        this.onProjectLongClickedListener = onProjectLongClickedListener;
    }

    public void submitList(@NonNull List<JavaProject> projects) {
         var diffResult = DiffUtil.calculateDiff(new DiffUtil.Callback() {
            @Override
            public int getOldListSize() {
                 return mProjects.size();
            }

            @Override
            public int getNewListSize() {
                return projects.size();
            }

            @Override
            public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
                return mProjects.get(oldItemPosition).equals(projects.get(newItemPosition));
            }

            @Override
            public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
                 return mProjects.get(oldItemPosition).equals(projects.get(newItemPosition));
            }
         });
         mProjects.clear();
         mProjects.addAll(projects);
         diffResult.dispatchUpdatesTo(this);
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        var root = LayoutInflater.from(parent.getContext()).inflate(R.layout.project_item, parent, false);
        final var holder = new ViewHolder(root);
        root.setOnClickListener(v -> {
            if (onProjectSelectedListener != null) {
                int position = holder.getBindingAdapterPosition();
                if (position != RecyclerView.NO_POSITION) {
                    onProjectSelectedListener.onProjectSelect(mProjects.get(position));
                }
            }
        });
        return holder;
    }

    @NonNull
    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        holder.bind(mProjects.get(position));
    }
    
    @Override
    public int getItemCount() {
        return mProjects.size(); 
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {

        private final TextView title;
        private final TextView path;

        public ViewHolder(View view) {
            super(view);
            title = view.findViewById(R.id.project_title);
            path = view.findViewById(R.id.project_path);
        }

        public void bind(JavaProject project) {
             title.setText(project.getProjectName());
             path.setText(project.getProjectDirPath());
        }
    }
}

File: app/src/main/java/org/cosmicide/rewrite/adapter/ProjectAdapter.kt
package org.cosmicide.rewrite.adapter

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import org.cosmicide.project.Project
import org.cosmicide.rewrite.databinding.ProjectItemBinding
import kotlin.properties.Delegates

/**
 * Adapter for displaying a list of [Project] objects in a [RecyclerView].
 *
 * @param listener Listener for handling user interactions with the items in the list.
 */
class ProjectAdapter(private val listener: OnProjectEventListener) : RecyclerView.Adapter<ProjectAdapter.ViewHolder>() {

    /**
     * List of [Project] objects to be displayed in the [RecyclerView].
     */
    private var projects: List<Project> by Delegates.observable(emptyList()) { _, oldList, newList ->
        DiffUtil.calculateDiff(object : DiffUtil.Callback() {
            override fun getOldListSize() = oldList.size
            override fun getNewListSize() = newList.size
            override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int) =
                oldList[oldItemPosition] == newList[newItemPosition]
            override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int) =
                oldList[oldItemPosition] == newList[newItemPosition]
        }).dispatchUpdatesTo(this)
    }

    /**
     * Interface for handling user interactions with the items in the list.
     */
    interface OnProjectEventListener {
        fun onProjectClicked(project: Project)
        fun onProjectLongClicked(project: Project): Boolean
    }

    /**
     * Sets the list of projects to be displayed in the [RecyclerView].
     *
     * @param projects The list of [Project] objects.
     */
    fun submitList(projects: List<Project>) {
        this.projects = projects
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder =
        ViewHolder(
            ProjectItemBinding.inflate(LayoutInflater.from(parent.context), parent, false),
            listener
        )

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bind(projects[position])
    }

    override fun getItemCount() = projects.size

    /**
     * ViewHolder for displaying a single [Project] object in the [RecyclerView].
     *
     * @param binding The [ProjectItemBinding] object for the item view.
     * @param listener The [OnProjectEventListener] for handling user interactions with the item view.
     */
    class ViewHolder(
        private val binding: ProjectItemBinding,
        private val listener: OnProjectEventListener
    ) : RecyclerView.ViewHolder(binding.root) {

        /**
         * Binds the given [Project] object to the item view.
         *
         * @param project The [Project] object to be displayed.
         */
        fun bind(project: Project) {
            binding.projectTitle.text = project.name
            binding.projectPath.text = project.root.absolutePath
            binding.root.setOnClickListener { listener.onProjectClicked(project) }
            binding.root.setOnLongClickListener { listener.onProjectLongClicked(project) }
        }
    }
}